Fixes #13164 - adding view_params permission

orrabin opened the pull request.

orrabin committed:
Fixes #13164 - adding view_params permission

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -393,6 +393,9 @@
     map.permission :edit_classes,   {:host_editing => [:edit_classes],
                                      :"api/v2/host_classes" => [:index, :create, :destroy]
                                     }
+    map.permission :view_params,  {:host_editing => [:view_params],
```
I don't see how this particular route is used.  Is the patch missing corresponding code in the view?

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -393,6 +393,9 @@
                                      :"api/v2/host_classes" => [:index, :create, :destroy]
                                     }
+    map.permission :view_params,  {:host_editing => [:view_params],
+                                     :"api/v2/parameters" => [:show]
```
This permission's granted through _some_ of the corresponding nested objects, e.g. view_domains/edit_domains.  Should it be migrated entirely, or removed from here and added to all nested objects?

domcleal commented:
It's unclear to me where this permission's used, if you can expand on that it would be appreciated.

orrabin commented:
@domcleal it's not currently in use, I was thinking of adding it in host/hostgroup edit form: https://github.com/orrabin/foreman/commit/f8929556e1428a8036ccc7fafe7d349c6e5f22b1
but that means doing the same for the rest of the parameters too.

orrabin commented:
closing this until I get more information from the bug reporter.

orrabin commented:
@domcleal this is still missing tests but I reopened with the migration and the views. I put the permission under `host_editing` and used that in all the view. Would it be better to put it in each nested object instead?

domcleal commented:
> I put the permission under host_editing and used that in all the view. Would it be better to put it in each nested object instead?

Yes, I think it probably would be good, but that's quite a change - it's also going to affect the permissions required to access different APIs significantly.  I'm happy to take this as-is or go for more granular permissions (or one, then the other).

Note the failing tests are on _upgrades_ from a stable branch, not on a clean database.

orrabin commented:
@domcleal I'll leave it as is for now. I added tests for parameters and fixed the migrations so updates won't fail.

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -394,18 +394,21 @@
   permission_set.security_block :host_editing do |map|
-    map.permission :edit_classes,   {:host_editing => [:edit_classes],
+    map.permission :edit_classes, {:host_editing => [:edit_classes],
                                      :"api/v2/host_classes" => [:index, :create, :destroy]
```
Nitpick, the symbols on the second lines aren't aligned to the first (`:host_editing` vs `:"api...`)

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -394,18 +394,21 @@
+    map.permission :view_params, {:host_editing => [:view_params],
+                                     :"api/v2/parameters" => [:show]
+                                 }
+    map.permission :create_params,{:host_editing => [:create_params],
```
Nitpick, missing space between permission name and hash

domcleal commented on db/migrate/20160317070258_add_view_params_to_filters_with_edit.rb:
```diff
@@ -0,0 +1,18 @@
+  def up
+    filters_to_update = Filter.joins(:permissions).where("permissions.name IN ('edit_params', 'create_params', 'destroy_params')").uniq
+    view_params = Permission.where(:name => 'view_params', :resource_type => 'Parameter').first_or_create
+    if view_params.present?
```
always true?

domcleal commented on test/functional/operatingsystems_controller_test.rb:
```diff
@@ -1,8 +1,12 @@
-    @request.session[:user] = users(:one).id
-    users(:one).roles       = [Role.default, Role.find_by_name('Viewer')]
+  def setup_user(operation = nil, type = 'operatingsystems')
+    if operation.nil?
```
It might be better to rename this OS-specific helper so it doesn't override the generic one - it'll be easier to see where it's used and remove it in future that way.

domcleal commented:
The DB migration might also need to add the view_params permission to the Viewer role if it exists, or perhaps preferably for future use, adapt the seed (03-roles) to modify the role and add the permission(s) if missing.

domcleal commented:
Oh, and could you also add the same test to the org/loc functionals? That's what the bug report at http://projects.theforeman.org/issues/13164#note-3 is describing.

orrabin commented:
@domcleal I added tests for org/loc. The tests passed for me as soon as I added them so the issue in the bug report should be fixed now.

For the Viewer role I only added it to the migration because the seeds file searches for every permission with 'view' in it so it should pick up the new permission.

orrabin commented:
The migrations are not failing locally for me, I'll check what's going on.
Edit: migration failed only if this is the first time they run so to reproduce I needed a clean db. It now works locally on an empty db.

ohadlevy commented:
@orrabin anything i can help to move this along? thanks!

orrabin commented:
@ohadlevy I rebased this today because of merge conflicts, it now can be reviewed again.
@domcleal I fixed your comments.
Test failure looks unrelated.

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -410,18 +410,21 @@
+                                    :"api/v2/host_classes" => [:index, :create, :destroy]
+                                }
+    map.permission :view_params, { :host_editing => [:view_params],
+                                   :"api/v2/parameters" => [:show]
```
missing :index

domcleal commented:
1) The view_domains permission should have api/v2/parameters removed, as it should now require view_params instead.

2) Probably the same for edit_domains and edit_hostgroups for full consistency.

ares commented on app/views/hostgroups/_form.html.erb:
```diff
@@ -64,15 +66,17 @@
+          <h2><%= _('Global parameters') %></h2>
+          <% if @hostgroup.parent.present? %>
+              <h4><%= _('Parent parameters') %></h4>
+              <%= render "common_parameters/inherited_parameters", { :inherited_parameters => @hostgroup.parent_params(true), :parameters => @hostgroup.group_parameters } %>
```
Since Parameter filters can be granular, I don't think `authorized_for` check is enough, user might be limited to just some parameters. `@hostgroup.parent_params` does not seem to solve it either. I think `.authorized(:view_params)` scope should be used at every place where Parameters are loaded. The same applies to other places of this PR.

orrabin committed:
added .authorized to parameters

orrabin commented:
closing in favor of: https://github.com/theforeman/foreman/pull/3592
